[BUG] mongodb 4/5 dump backup failed

**Describe the bug**
A clear and concise description of what the bug is.

**To Reproduce**
Steps to reproduce the behavior:
1. create cluster
```
apiVersion: apps.kubeblocks.io/v1
kind: Cluster
metadata:
  name: mongodbs-yccyjr
  namespace: default
spec:
  clusterDef: mongodb
  topology: sharding
  terminationPolicy: DoNotTerminate
  shardings:
    - name: shard
      shards: 3
      template:
        name: mongodb
        serviceVersion: 5.0.29
        replicas: 3
        disableExporter: false
        env:
          - name: SHARDING_NAME
            value: shard
        resources:
          requests:
            cpu: 200m
            memory: 0.5Gi
          limits:
            cpu: 200m
            memory: 0.5Gi
        persistentVolumeClaimRetentionPolicy:
          whenDeleted: Delete
        volumeClaimTemplates:
          - name: data
            spec:
              storageClassName: 
              accessModes:
                - ReadWriteOnce
              resources:
                requests:
                  storage: 20Gi
  componentSpecs:
    - name: config-server
      serviceVersion: 5.0.29
      replicas: 3
      disableExporter: false
      systemAccounts:
      - name: root
        passwordConfig:
          length: 16
          numDigits: 8
          numSymbols: 0
          letterCase: MixedCases
          seed: mongodbs-yccyjr
      env:
        - name: MONGODB_BALANCER_ENABLED
          value: "false"
      resources:
        limits:
          cpu: 200m
          memory: 0.5Gi
        requests:
          cpu: 200m
          memory: 0.5Gi
      persistentVolumeClaimRetentionPolicy:
        whenDeleted: Delete
      volumeClaimTemplates:
        - name: data
          spec:
            storageClassName: 
            accessModes:
              - ReadWriteOnce
            resources:
              requests:
                storage: 20Gi
    - name: mongos
      serviceVersion: 5.0.29
      replicas: 3
      disableExporter: false
      env:
        - name: MONGODB_BALANCER_ENABLED
          value: "false"
      resources:
        limits:
          cpu: 200m
          memory: 0.5Gi
        requests:
          cpu: 200m
          memory: 0.5Gi
``` 

```
      `kbcli cluster list mongodbs-yccyjr --show-labels  --namespace default `
    
NAME              NAMESPACE   CLUSTER-DEFINITION   TERMINATION-POLICY   STATUS     CREATED-TIME                 LABELS                                         
mongodbs-yccyjr   default     mongodb              DoNotTerminate       Creating   Aug 13,2026 15:40 UTC+0800   

      `kbcli cluster list-instances mongodbs-yccyjr --namespace default `
    
NAME                              NAMESPACE   CLUSTER           COMPONENT          STATUS    ROLE        ACCESSMODE   AZ               CPU(REQUEST/LIMIT)   MEMORY(REQUEST/LIMIT)   STORAGE     NODE                          CREATED-TIME                 
mongodbs-yccyjr-config-server-0   default     mongodbs-yccyjr   config-server      Running   primary                  cn-guangzhou-b   200m / 200m          512Mi / 512Mi           data:20Gi   192.168.0.35/192.168.0.35     Aug 13,2026 15:40 UTC+0800   
mongodbs-yccyjr-config-server-1   default     mongodbs-yccyjr   config-server      Running   secondary                cn-guangzhou-b   200m / 200m          512Mi / 512Mi           data:20Gi   192.168.0.134/192.168.0.134   Aug 13,2026 15:40 UTC+0800   
mongodbs-yccyjr-config-server-2   default     mongodbs-yccyjr   config-server      Running   secondary                cn-guangzhou-b   200m / 200m          512Mi / 512Mi           data:20Gi   192.168.0.133/192.168.0.133   Aug 13,2026 15:40 UTC+0800   
mongodbs-yccyjr-mongos-0          default     mongodbs-yccyjr   mongos             Running   <none>                   cn-guangzhou-b   200m / 200m          512Mi / 512Mi           <none>      192.168.0.35/192.168.0.35     Aug 13,2026 15:40 UTC+0800   
mongodbs-yccyjr-mongos-1          default     mongodbs-yccyjr   mongos             Running   <none>                   cn-guangzhou-b   200m / 200m          512Mi / 512Mi           <none>      192.168.0.134/192.168.0.134   Aug 13,2026 15:40 UTC+0800   
mongodbs-yccyjr-mongos-2          default     mongodbs-yccyjr   mongos             Running   <none>                   cn-guangzhou-b   200m / 200m          512Mi / 512Mi           <none>      192.168.0.133/192.168.0.133   Aug 13,2026 15:40 UTC+0800   
mongodbs-yccyjr-shard-f47-0       default     mongodbs-yccyjr   shard(shard-f47)   Running   primary                  cn-guangzhou-b   200m / 200m          512Mi / 512Mi           data:20Gi   192.168.0.133/192.168.0.133   Aug 13,2026 15:40 UTC+0800   
mongodbs-yccyjr-shard-f47-1       default     mongodbs-yccyjr   shard(shard-f47)   Running   secondary                cn-guangzhou-b   200m / 200m          512Mi / 512Mi           data:20Gi   192.168.0.134/192.168.0.134   Aug 13,2026 15:40 UTC+0800   
mongodbs-yccyjr-shard-f47-2       default     mongodbs-yccyjr   shard(shard-f47)   Running   secondary                cn-guangzhou-b   200m / 200m          512Mi / 512Mi           data:20Gi   192.168.0.35/192.168.0.35     Aug 13,2026 15:40 UTC+0800   
mongodbs-yccyjr-shard-nsw-0       default     mongodbs-yccyjr   shard(shard-nsw)   Running   primary                  cn-guangzhou-b   200m / 200m          512Mi / 512Mi           data:20Gi   192.168.0.134/192.168.0.134   Aug 13,2026 15:40 UTC+0800   
mongodbs-yccyjr-shard-nsw-1       default     mongodbs-yccyjr   shard(shard-nsw)   Running   secondary                cn-guangzhou-b   200m / 200m          512Mi / 512Mi           data:20Gi   192.168.0.118/192.168.0.118   Aug 13,2026 15:40 UTC+0800   
mongodbs-yccyjr-shard-nsw-2       default     mongodbs-yccyjr   shard(shard-nsw)   Running   secondary                cn-guangzhou-b   200m / 200m          512Mi / 512Mi           data:20Gi   192.168.0.35/192.168.0.35     Aug 13,2026 15:40 UTC+0800   
mongodbs-yccyjr-shard-v49-0       default     mongodbs-yccyjr   shard(shard-v49)   Running   primary                  cn-guangzhou-b   200m / 200m          512Mi / 512Mi           data:20Gi   192.168.0.133/192.168.0.133   Aug 13,2026 15:40 UTC+0800   
mongodbs-yccyjr-shard-v49-1       default     mongodbs-yccyjr   shard(shard-v49)   Running   secondary                cn-guangzhou-b   200m / 200m          512Mi / 512Mi           data:20Gi   192.168.0.118/192.168.0.118   Aug 13,2026 15:40 UTC+0800   
mongodbs-yccyjr-shard-v49-2       default     mongodbs-yccyjr   shard(shard-v49)   Running   secondary                cn-guangzhou-b   200m / 200m          512Mi / 512Mi           data:20Gi   192.168.0.134/192.168.0.134   Aug 13,2026 15:40 UTC+0800   

```
2. dump backup
```
kbcli cluster backup mongodbs-yccyjr --method dump --namespace defaul
```
3. See error
```
kubectl get backup  backup-default-mongodbs-yccyjr-20260813154129
NAME                                            POLICY                                        METHOD   REPO                    STATUS   TOTAL-SIZE   DURATION   DELETION-POLICY   CREATION-TIME          COMPLETION-TIME   EXPIRATION-TIME
backup-default-mongodbs-yccyjr-20260813154129   mongodbs-yccyjr-config-server-backup-policy   dump     backuprepo-kbcli-test   Failed                           Delete            2026-08-13T07:41:30Z                     

```
describe backup
```
kubectl describe backup backup-default-mongodbs-yccyjr-20260813154129 
Name:         backup-default-mongodbs-yccyjr-20260813154129
Namespace:    default
Labels:       app.kubernetes.io/instance=mongodbs-yccyjr
              app.kubernetes.io/managed-by=kubeblocks-dataprotection
              apps.kubeblocks.io/component-name=config-server
              dataprotection.kubeblocks.io/backup-policy=mongodbs-yccyjr-config-server-backup-policy
              dataprotection.kubeblocks.io/backup-repo-name=backuprepo-kbcli-test
              dataprotection.kubeblocks.io/backup-type=Full
              dataprotection.kubeblocks.io/cluster-uid=3ecff161-5283-4a87-8565-fb4b7fbb1d58
              operations.kubeblocks.io/ops-name=backup-default-mongodbs-yccyjr-20260813154129
              operations.kubeblocks.io/ops-type=Backup
Annotations:  kubeblocks.io/cluster-snapshot:
                {"metadata":{"name":"mongodbs-yccyjr","namespace":"default","creationTimestamp":null,"annotations":{"kubeblocks.io/crd-api-version":"apps....
              kubeblocks.io/encrypted-system-accounts: {"config-server":{"root":"qHgfElZb8RM+nIEoAyF7s7eF6iGtsp8J8ZwWjB+QxNKNbI9AOzUxWNBNfC4="}}
API Version:  dataprotection.kubeblocks.io/v1alpha1
Kind:         Backup
Metadata:
  Creation Timestamp:  2026-08-13T07:41:30Z
  Finalizers:
    dataprotection.kubeblocks.io/finalizer
  Generation:        1
  Resource Version:  267029105
  UID:               de441b79-6ce5-4495-aa5b-a8f76e551445
Spec:
  Backup Method:       dump
  Backup Policy Name:  mongodbs-yccyjr-config-server-backup-policy
  Deletion Policy:     Delete
Status:
  Actions:
    Action Type:           Job
    Completion Timestamp:  2026-08-13T07:43:36Z
    Name:                  dp-prebackup-0-0
    Object Ref:
      API Version:         batch/v1
      Kind:                Job
      Name:                dp-prebackup-0-0-backup-default-mongodbs-yccyjr-20260813154129
      Namespace:           default
      Resource Version:    267024805
      UID:                 751af7c3-0657-4cde-b83c-e15c5ed9c21e
    Phase:                 Completed
    Start Timestamp:       2026-08-13T07:41:30Z
    Target Pod Name:       mongodbs-yccyjr-config-server-1
    Action Type:           Job
    Completion Timestamp:  2026-08-13T07:43:36Z
    Failure Reason:        Log Collector Output: for the oplog
2026-08-13T07:43:24Z I [mongodbs-yccyjr-config-server/mongodbs-yccyjr-config-server-1.mongodbs-yccyjr-config-server-headless.default.svc.cluster.local:27017] [backup/2026-08-13T07:43:19Z] mongodump finished, waiting for the oplog
2026-08-13T07:43:26Z I [mongodbs-yccyjr-config-server/mongodbs-yccyjr-config-server-1.mongodbs-yccyjr-config-server-headless.default.svc.cluster.local:27017] [backup/2026-08-13T07:43:19Z] created chunk 2026-08-13T07:43:20 - 2026-08-13T07:43:25
failed with exit code 1

    Name:  dp-backup-0
    Object Ref:
      API Version:       batch/v1
      Kind:              Job
      Name:              dp-backup-0-backup-default-mongodbs-yccyjr-20260813154129-de441
      Namespace:         default
      Resource Version:  267029098
      UID:               a7c9b472-3e05-4323-8f31-020704d29476
    Phase:               Failed
    Start Timestamp:     2026-08-13T07:41:34Z
    Target Pod Name:     mongodbs-yccyjr-config-server-1
  Backup Method:
    Action Set Name:  mongodb-shard-pbm-logical
    Env:
      Name:            PBM_IMAGE_TAG
      Value:           2.5.0-multi
      Name:            PSM_IMAGE_TAG
      Value:           5.0.29-multi
    Name:              dump
    Snapshot Volumes:  false
    Target Volumes:
      Volume Mounts:
        Mount Path:  /data/mongodb
        Name:        data
  Backup Repo Name:  backuprepo-kbcli-test
  Failure Reason:    there are failed actions, you can obtain the more information in the status.actions
  Format Version:    0.1.0
  Path:              /default/mongodbs-yccyjr-3ecff161-5283-4a87-8565-fb4b7fbb1d58/config-server/backup-default-mongodbs-yccyjr-20260813154129
  Phase:             Failed
  Start Timestamp:   2026-08-13T07:41:30Z
  Target:
    Connection Credential:
      Password Key:  password
      Secret Name:   mongodbs-yccyjr-config-server-account-root
      Username Key:  username
    Pod Selector:
      Fallback Label Selector:
        Match Labels:
          app.kubernetes.io/instance:         mongodbs-yccyjr
          app.kubernetes.io/managed-by:       kubeblocks
          apps.kubeblocks.io/component-name:  config-server
          kubeblocks.io/role:                 primary
      Match Labels:
        app.kubernetes.io/instance:         mongodbs-yccyjr
        app.kubernetes.io/managed-by:       kubeblocks
        apps.kubeblocks.io/component-name:  config-server
        kubeblocks.io/role:                 secondary
      Strategy:                             Any
    Selected Target Pods:
      mongodbs-yccyjr-config-server-1
Events:
  Type     Reason           Age                From               Message
  ----     ------           ----               ----               -------
  Normal   CreatingJob      17m (x2 over 17m)  backup-controller  creating job default/dp-prebackup-0-0-backup-default-mongodbs-yccyjr-20260813154129
  Normal   CreatingJob      17m (x2 over 17m)  backup-controller  creating job default/dp-backup-0-backup-default-mongodbs-yccyjr-20260813154129-de441
  Warning  ReconcileFailed  15m                backup-controller  Reconciling failed, error: there are failed actions, you can obtain the more information in the status.actions
```
```
kubectl get pod -l app.kubernetes.io/instance=mongodbs-yccyjr 
NAME                                                              READY   STATUS      RESTARTS   AGE
dp-backup-0-backup-default-mongodbs-yccyjr-20260813154129-jhkj5   0/2     Error       0          3m45s
dp-backup-0-backup-default-mongodbs-yccyjr-20260813154129-z5kqk   0/2     Error       0          3m8s
dp-backup-0-backup-default-mongodbs-yccyjr-20260813154129-zprdb   0/2     Error       0          4m33s
dp-prebackup-0-0-backup-default-mongodbs-yccyjr-2026081315skxzp   0/1     Completed   0          4m33s
mongodbs-yccyjr-config-server-0                                   4/4     Running     0          6m16s
mongodbs-yccyjr-config-server-1                                   4/4     Running     0          6m17s
mongodbs-yccyjr-config-server-2                                   4/4     Running     0          6m17s
mongodbs-yccyjr-mongos-0                                          2/2     Running     0          6m17s
mongodbs-yccyjr-mongos-1                                          2/2     Running     0          6m17s
mongodbs-yccyjr-mongos-2                                          2/2     Running     0          6m17s
mongodbs-yccyjr-shard-f47-0                                       4/4     Running     0          6m17s
mongodbs-yccyjr-shard-f47-1                                       4/4     Running     0          6m17s
mongodbs-yccyjr-shard-f47-2                                       4/4     Running     0          6m17s
mongodbs-yccyjr-shard-nsw-0                                       4/4     Running     0          6m16s
mongodbs-yccyjr-shard-nsw-1                                       4/4     Running     0          6m17s
mongodbs-yccyjr-shard-nsw-2                                       4/4     Running     0          6m17s
mongodbs-yccyjr-shard-v49-0                                       4/4     Running     0          6m17s
mongodbs-yccyjr-shard-v49-1                                       4/4     Running     0          6m17s
mongodbs-yccyjr-shard-v49-2                                       4/4     Running     0          6m16s
```
describe error pod 
```
kubectl describe pod dp-backup-0-backup-default-mongodbs-yccyjr-20260813154129-zprdb
Name:             dp-backup-0-backup-default-mongodbs-yccyjr-20260813154129-zprdb
Namespace:        default
Priority:         0
Service Account:  kubeblocks-dataprotection-worker
Node:             192.168.0.134/192.168.0.134
Start Time:       Thu, 13 Aug 2026 15:41:52 +0800
Labels:           app.kubernetes.io/instance=mongodbs-yccyjr
                  app.kubernetes.io/managed-by=kubeblocks-dataprotection
                  batch.kubernetes.io/controller-uid=a7c9b472-3e05-4323-8f31-020704d29476
                  batch.kubernetes.io/job-name=dp-backup-0-backup-default-mongodbs-yccyjr-20260813154129-de441
                  controller-uid=a7c9b472-3e05-4323-8f31-020704d29476
                  dataprotection.kubeblocks.io/backup-name=backup-default-mongodbs-yccyjr-20260813154129
                  dataprotection.kubeblocks.io/backup-policy=mongodbs-yccyjr-config-server-backup-policy
                  dataprotection.kubeblocks.io/backup-repo-name=backuprepo-kbcli-test
                  dataprotection.kubeblocks.io/backup-type=Full
                  dataprotection.kubeblocks.io/cluster-uid=3ecff161-5283-4a87-8565-fb4b7fbb1d58
                  job-name=dp-backup-0-backup-default-mongodbs-yccyjr-20260813154129-de441
                  operations.kubeblocks.io/ops-name=backup-default-mongodbs-yccyjr-20260813154129
                  operations.kubeblocks.io/ops-type=Backup
Annotations:      vke.volcengine.com/cello-pod-evict-policy: allow
Status:           Failed
IP:               192.168.0.254
IPs:
  IP:           192.168.0.254
Controlled By:  Job/dp-backup-0-backup-default-mongodbs-yccyjr-20260813154129-de441
Init Containers:
  dp-copy-datasafed:
    Container ID:  containerd://e0955beb1619ff6a2fd5343d2e8f53662b448a10d9ce41f298e27a57c3de574c
    Image:         apecloud-registry.cn-zhangjiakou.cr.aliyuncs.com/apecloud/datasafed:0.2.3
    Image ID:      apecloud-registry.cn-zhangjiakou.cr.aliyuncs.com/apecloud/datasafed@sha256:7775e8184fbc833ee089b33427c4981bd7cd7d98cce5aeff1a9856b5de966b0f
    Port:          <none>
    Host Port:     <none>
    Command:
      /bin/sh
      -c
      /scripts/install-datasafed.sh /bin/datasafed
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Thu, 13 Aug 2026 15:41:53 +0800
      Finished:     Thu, 13 Aug 2026 15:41:53 +0800
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     0
      memory:  0
    Requests:
      cpu:        0
      memory:     0
    Environment:  <none>
    Mounts:
      /bin/datasafed from dp-datasafed-bin (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-wfmhj (ro)
Containers:
  backupdata:
    Container ID:  containerd://ebf60f77e8f11aaf598d7c33fab6f509a1582ad340e2dea1ce79f84205791c5b
    Image:         apecloud-registry.cn-zhangjiakou.cr.aliyuncs.com/apecloud/percona-backup-mongodb:2.5.0-multi
    Image ID:      apecloud-registry.cn-zhangjiakou.cr.aliyuncs.com/apecloud/percona-backup-mongodb@sha256:4c9decaa78363aae9d815303f47c7c77936144f37844a29c87b16d87df9f1557
    Port:          <none>
    Host Port:     <none>
    Command:
      bash
      -c
      #!/bin/bash
      # Copyright ApeCloud Co., Ltd. All Rights Reserved.
      # SPDX-License-Identifier: LicenseRef-KubeBlocks-Enterprise
      # shellcheck disable=SC2086
      
      function handle_exit() {
        exit_code=$?
        if [ $exit_code -ne 0 ]; then
          echo "failed with exit code $exit_code"
          touch "${DP_BACKUP_INFO_FILE}.exit"
          exit 1
        fi
      }
      # log info file
      function DP_log() {
          msg=$1
          local curr_date=$(date -u '+%Y-%m-%d %H:%M:%S')
          echo "${curr_date} INFO: $msg"
      }
      
      # log error info
      function DP_error_log() {
          msg=$1
          local curr_date=$(date -u '+%Y-%m-%d %H:%M:%S')
          echo "${curr_date} ERROR: $msg"
      }
      
      function buildJsonString() {
          local jsonString=${1}
          local key=${2}
          local value=${3}
          if [ ! -z "$jsonString" ];then
             jsonString="${jsonString},"
          fi
          echo "${jsonString}\"${key}\":\"${value}\""
      }
      
      # Save backup status info file for syncing progress.
      # timeFormat: %Y-%m-%dT%H:%M:%SZ
      function DP_save_backup_status_info() {
          export PATH="$PATH:$DP_DATASAFED_BIN_PATH"
          export DATASAFED_BACKEND_BASE_PATH="$DP_BACKUP_BASE_PATH"
      
          local totalSize=$1
          local startTime=$2
          local stopTime=$3
          local timeZone=$4
          local extras=$5
          local timeZoneStr=""
          if [ ! -z ${timeZone} ]; then
             timeZoneStr=",\"timeZone\":\"${timeZone}\""
          fi
          if [ -z "${stopTime}" ];then
            echo "{\"totalSize\":\"${totalSize}\"}" > ${DP_BACKUP_INFO_FILE}
          elif [ -z "${startTime}" ];then
            echo "{\"totalSize\":\"${totalSize}\",\"extras\":[${extras}],\"timeRange\":{\"end\":\"${stopTime}\"${timeZoneStr}}}" > ${DP_BACKUP_INFO_FILE}
          else
            echo "{\"totalSize\":\"${totalSize}\",\"extras\":[${extras}],\"timeRange\":{\"start\":\"${startTime}\",\"end\":\"${stopTime}\"${timeZoneStr}}}" > ${DP_BACKUP_INFO_FILE}
          fi
      }
      
      function getToolConfigValue() {
        local var=$1
        cat "$toolConfig" | grep "$var" | awk '{print $NF}'
      }
      
      function set_backup_config_env() {
        toolConfig=/etc/datasafed/datasafed.conf
        if [ ! -f ${toolConfig} ]; then
          DP_error_log "Config file not found: ${toolConfig}"
          exit 1
        fi
      
        local provider=""
        local access_key_id=""
        local secret_access_key=""
        local region=""
        local endpoint=""
        local bucket=""
      
        IFS=$'\n'
        for line in $(cat ${toolConfig}); do
          line=$(eval echo $line)
          if [[ $line == "access_key_id"* ]]; then
            access_key_id=$(getToolConfigValue "$line")
          elif [[ $line == "secret_access_key"* ]]; then
            secret_access_key=$(getToolConfigValue "$line")
          elif [[ $line == "region"* ]]; then
            region=$(getToolConfigValue "$line")
          elif [[ $line == "endpoint"* ]]; then
            endpoint=$(getToolConfigValue "$line")
          elif [[ $line == "root"* ]]; then
            bucket=$(getToolConfigValue "$line")
          elif [[ $line == "provider"* ]]; then
            provider=$(getToolConfigValue "$line")
          fi
        done
      
        if [[ ! $endpoint =~ ^https?:// ]]; then
          endpoint="https://${endpoint}"
        fi
      
        if [[ "$provider" == "Alibaba" ]]; then
          regex='https?:\/\/oss-(.*?)\.aliyuncs\.com'
          if [[ "$endpoint" =~ $regex ]]; then
            region="${BASH_REMATCH[1]}"
            DP_log "Extract region from $endpoint-> $region"
          else
            DP_log "Failed to extract region from endpoint: $endpoint"
          fi
        elif [[ "$provider" == "TencentCOS" ]]; then
          regex='https?:\/\/cos\.(.*?)\.myqcloud\.com'
          if [[ "$endpoint" =~ $regex ]]; then
            region="${BASH_REMATCH[1]}"
            DP_log "Extract region from $endpoint-> $region"
          else
            DP_log "Failed to extract region from endpoint: $endpoint"
          fi
        elif [[ "$provider" == "Minio" ]]; then
          export S3_FORCE_PATH_STYLE="true"
        else
          echo "Unsupported provider: $provider"
        fi
        backup_path=$(dirname "$DP_BACKUP_BASE_PATH")
      
        export S3_ACCESS_KEY="${access_key_id}"
        export S3_SECRET_KEY="${secret_access_key}"
        export S3_REGION="${region}"
        export S3_ENDPOINT="${endpoint}"
        export S3_BUCKET="${bucket}"
        export S3_PREFIX="${backup_path#/}/$PBM_BACKUP_DIR_NAME"
      
        DP_log "storage config have been extracted."
      }
      
      # config backup agent
      generate_endpoints() {
          local fqdns=$1
          local port=$2
      
          if [ -z "$fqdns" ]; then
              echo "ERROR: No FQDNs provided for endpoints." >&2
              exit 1
          fi
      
          IFS=',' read -ra fqdn_array <<< "$fqdns"
          local endpoints=()
      
          for fqdn in "${fqdn_array[@]}"; do
              trimmed_fqdn=$(echo "$fqdn" | xargs)
              if [[ -n "$trimmed_fqdn" ]]; then
                  endpoints+=("${trimmed_fqdn}:${port}")
              fi
          done
      
          IFS=','; echo "${endpoints[*]}"
      }
      
      function export_pbm_env_vars() {
        export PBM_AGENT_MONGODB_USERNAME="$MONGODB_USER"
        export PBM_AGENT_MONGODB_PASSWORD="$MONGODB_PASSWORD"
      
        cfg_server_endpoints="$(generate_endpoints "$CFG_SERVER_POD_FQDN_LIST" "$CFG_SERVER_INTERNAL_PORT")"
        export PBM_MONGODB_URI="mongodb://$PBM_AGENT_MONGODB_USERNAME:$PBM_AGENT_MONGODB_PASSWORD@$cfg_server_endpoints/?authSource=admin&replSetName=$CFG_SERVER_REPLICA_SET_NAME"
      }
      
      function export_pbm_env_vars_for_rs() {
        export PBM_AGENT_MONGODB_USERNAME="$MONGODB_USER"
        export PBM_AGENT_MONGODB_PASSWORD="$MONGODB_PASSWORD"
      
        mongodb_endpoints="$(generate_endpoints "$MONGODB_POD_FQDN_LIST" "$KB_SERVICE_PORT")"
        export PBM_MONGODB_URI="mongodb://$PBM_AGENT_MONGODB_USERNAME:$PBM_AGENT_MONGODB_PASSWORD@$mongodb_endpoints/?authSource=admin&replSetName=$MONGODB_REPLICA_SET_NAME"
      }
      
      function sync_pbm_storage_config() {
        echo "INFO: Checking if PBM storage config exists"
        pbm_config_exists=true
        check_config=$(pbm config --mongodb-uri "$PBM_MONGODB_URI" -o json) || {
          pbm_config_exists=false
          echo "INFO: PBM storage config does not exist."
        }
        if [ "$pbm_config_exists" = "true" ]; then
          # check_config=$(pbm config --mongodb-uri "$PBM_MONGODB_URI" -o json)
          current_endpoint=$(echo "$check_config" | jq -r '.storage.s3.endpointUrl')
          current_region=$(echo "$check_config" | jq -r '.storage.s3.region')
          current_bucket=$(echo "$check_config" | jq -r '.storage.s3.bucket')
          current_prefix=$(echo "$check_config" | jq -r '.storage.s3.prefix')
          echo "INFO: Current PBM storage endpoint: $current_endpoint"
          echo "INFO: Current PBM storage region: $current_region"
          echo "INFO: Current PBM storage bucket: $current_bucket"
          echo "INFO: Current PBM storage prefix: $current_prefix"
          if [ "$current_prefix" = "$S3_PREFIX" ] && [ "$current_region" = "$S3_REGION" ] && [ "$current_bucket" = "$S3_BUCKET" ] && [ "$current_endpoint" = "$S3_ENDPOINT" ]; then
            echo "INFO: PBM storage config already exists."
          else
            pbm_config_exists=false
          fi
        fi
        if [ "$pbm_config_exists" = "false" ]; then
          cat <<EOF | pbm config --mongodb-uri "$PBM_MONGODB_URI" --file /dev/stdin > /dev/null
      storage:
        type: s3
        s3:
          region: ${S3_REGION}
          bucket: ${S3_BUCKET}
          prefix: ${S3_PREFIX}
          endpointUrl: ${S3_ENDPOINT}
          forcePathStyle: ${S3_FORCE_PATH_STYLE:-false}
          credentials:
            access-key-id: ${S3_ACCESS_KEY}
            secret-access-key: ${S3_SECRET_KEY}
      restore:
        numDownloadWorkers: ${PBM_RESTORE_DOWNLOAD_WORKERS:-4}
      backup:
        timeouts:
          startingStatus: 60
      EOF
          sleep 5
          echo "INFO: PBM storage configuration completed."
        fi
      }
      
      function print_pbm_logs_by_event() {
        local pbm_event=$1
        # echo "INFO: Printing PBM logs by event: $pbm_event"
        # shellcheck disable=SC2328
        local pbm_logs=$(pbm logs -e $pbm_event --tail 200 --mongodb-uri "$PBM_MONGODB_URI" > /dev/null)
        local purged_logs=$(echo "$pbm_logs" | awk -v start="$PBM_LOGS_START_TIME" '$1 >= start')
        if [ -z "$purged_logs" ]; then
          return
        fi
        echo "$purged_logs"
        # echo "INFO: PBM logs by event: $pbm_event printed."
      }
      
      function print_pbm_tail_logs() {
        echo "INFO: Printing PBM tail logs"
        pbm logs --tail 20 --mongodb-uri "$PBM_MONGODB_URI"
      }
      
      function handle_backup_exit() {
        exit_code=$?
        set +e
        if [ $exit_code -ne 0 ]; then
          print_pbm_tail_logs
      
          echo "failed with exit code $exit_code"
          touch "${DP_BACKUP_INFO_FILE}.exit"
          exit 1
        fi
      }
      
      function handle_restore_exit() {
        exit_code=$?
        set +e
        if [ $exit_code -ne 0 ]; then
          print_pbm_tail_logs
      
          echo "failed with exit code $exit_code"
          exit 1
        fi
      }
      
      function handle_pitr_exit() {
        exit_code=$?
        set +e
        if [[ "$PBM_DISABLE_PITR_WHEN_EXIT" == "true" ]]; then
          disable_pitr
        fi
      
        if [ $exit_code -ne 0 ]; then
          print_pbm_tail_logs
      
          echo "failed with exit code $exit_code"
          touch "${DP_BACKUP_INFO_FILE}.exit"
          exit 1
        fi
      }
      
      function wait_for_other_operations() {
        status_result=$(pbm status --mongodb-uri "$PBM_MONGODB_URI" -o json) || {
          echo "INFO: PBM is not configured."
          return
        }
        local except_type=$1
        local running_status=$(echo "$status_result" | jq -r '.running')
        local retry_count=0
        local max_retries=60
        while [ -n "$running_status" ] && [ "$running_status" != "{}" ] && [ $retry_count -lt $max_retries ]; do
          retry_count=$((retry_count+1))
          local running_type=$(echo "$running_status" | jq -r '.type')
          if [ -n "$running_type" ] && [ "$running_type" = "$except_type" ]; then
            break
          fi
          echo "INFO: Other operation $running_type is running, waiting... ($retry_count/$max_retries)"
          sleep 5
          running_status=$(pbm status --mongodb-uri "$PBM_MONGODB_URI" -o json | jq -r '.running')
        done
        if [ $retry_count -ge $max_retries ]; then
          echo "ERROR: Other operations are still running after $max_retries retries"
          exit 1
        fi
      }
      
      function export_logs_start_time_env() {
        local logs_start_time=$(date +"%Y-%m-%dT%H:%M:%SZ")
        export PBM_LOGS_START_TIME="${logs_start_time}"
      }
      
      function sync_pbm_config_from_storage() {
        echo "INFO: Syncing PBM config from storage..."
      
        wait_for_other_operations
      
        pbm config --force-resync --mongodb-uri "$PBM_MONGODB_URI"
        # print_pbm_logs_by_event "resync"
      
        # resync wait flag might don't work
        wait_for_other_operations
      
        echo "INFO: PBM config synced from storage."
      }
      
      function wait_for_backup_completion() {
        describe_result=""
        local retry_interval=5
        local attempt=1
        local max_retries=12
        set +e
        while true; do
          describe_result=$(pbm describe-backup --mongodb-uri "$PBM_MONGODB_URI" "$backup_name" -o json 2>&1)
          if [ $? -eq 0 ] && [ -n "$describe_result" ]; then
            backup_status=$(echo "$describe_result" | jq -r '.status')
            if [ "$backup_status" = "starting" ] || [ "$backup_status" = "running" ]; then
              echo "INFO: Backup status is $backup_status, retrying in ${retry_interval}s..."
            elif [ "$backup_status" = "" ]; then
              echo "INFO: Backup status is $backup_status, retrying in ${retry_interval}s..."
              attempt=$((attempt+1))
            elif [ "$backup_status" = "done" ]; then
              echo "INFO: Backup status is done."
              break
            else
              echo "ERROR: Backup failed with status: $backup_status"
              exit 1
            fi
          elif echo "$describe_result" | grep -q "not found"; then
            echo "INFO: Backup metadata not found, retrying in ${retry_interval}s..."
            attempt=$((attempt+1))
          else
            echo "ERROR: Unexpected: $describe_result"
            exit 1
          fi
          sleep $retry_interval
          if [ $attempt -gt $max_retries ]; then
            echo "ERROR: Failed to get backup status after $max_retries attempts"
            exit 1
          fi
        done
        set -e
      
        backup_status=$(echo "$describe_result" | jq -r '.status')
        if [ "$backup_status" != "done" ]; then
            echo "ERROR: Backup did not complete successfully, final status: $backup_status"
            exit 1
        fi
      }
      
      function create_restore_signal() {
          phase=$1
          kubectl apply -f - <<EOF
      apiVersion: v1
      kind: ConfigMap
      metadata:
        name: $dp_cm_name
        namespace: $dp_cm_namespace
        labels:
          app.kubernetes.io/instance: $CLUSTER_NAME
          apps.kubeblocks.io/restore-mongodb-shard: $phase
        ownerReferences:
          - apiVersion: apps.kubeblocks.io/v1
            blockOwnerDeletion: true
            controller: true
            kind: Cluster
            name: $CLUSTER_NAME
            uid: $CLUSTER_UID
      EOF
      }
      
      function process_restore_start_signal() {
          echo "INFO: Waiting for prepare restore start signal..."
          dp_cm_name="$CLUSTER_NAME-restore-signal"
          dp_cm_namespace="$CLUSTER_NAMESPACE"
          while true; do
              set +e
              kubectl_get_result=$(kubectl get configmap $dp_cm_name -n $dp_cm_namespace -o json 2>&1)
              kubectl_get_exit_code=$?
              set -e
              # Wait for the restore signal ConfigMap to be created or updated
              if [[ "$kubectl_get_exit_code" -ne 0 ]]; then
                  if [[ "$kubectl_get_result" == *"not found"* ]]; then
                      create_restore_signal "start"
                  fi
              else
                  annotation_value=$(echo "$kubectl_get_result" | jq -r '.metadata.labels["apps.kubeblocks.io/restore-mongodb-shard"] // empty')
                  if [[ "$annotation_value" == "start" ]]; then
                      break
                  elif [[ "$annotation_value" == "end" ]]; then
                      echo "INFO: Restore completed, exiting."
                      exit 0
                  else
                      echo "INFO: Restore start signal is $annotation_value, updating..."
                      create_restore_signal "start"
                  fi
              fi
              sleep 1
          done
          sleep 5
          echo "INFO: Prepare restore start signal completed."
      }
      
      function process_restore_end_signal() {
          echo "INFO: Waiting for prepare restore end signal..."
          sleep 5
          dp_cm_name="$CLUSTER_NAME-restore-signal"
          dp_cm_namespace="$CLUSTER_NAMESPACE"
          while true; do
              set +e
              kubectl_get_result=$(kubectl get configmap $dp_cm_name -n $dp_cm_namespace -o json 2>&1)
              kubectl_get_exit_code=$?
              set -e
              # Wait for the restore signal ConfigMap to be created or updated
              if [[ "$kubectl_get_exit_code" -ne 0 ]]; then
                  if [[ "$kubectl_get_result" == *"not found"* ]]; then
                      create_restore_signal "end"
                  fi
              else
                  annotation_value=$(echo "$kubectl_get_result" | jq -r '.metadata.labels["apps.kubeblocks.io/restore-mongodb-shard"] // empty')
                  if [[ "$annotation_value" == "end" ]]; then
                      break
                  else
                      echo "INFO: Restore end signal is $annotation_value, updating..."
                      create_restore_signal "end"
                  fi
              fi
              sleep 1
          done
          echo "INFO: Prepare restore end signal completed."
      }
      
      function get_describe_backup_info() {
        describe_result=""
        local max_retries=60
        local retry_interval=5
        local attempt=1
        set +e
        while [ $attempt -le $max_retries ]; do
            describe_result=$(pbm describe-backup --mongodb-uri "$PBM_MONGODB_URI" "$backup_name" -o json 2>&1)
            if [ $? -eq 0 ] && [ -n "$describe_result" ]; then
                break
            elif echo "$describe_result" | grep -q "not found"; then
                echo "INFO: Attempt $attempt: backup $backup_name not found, retrying in ${retry_interval}s..."
                if [ $((attempt % 30)) -eq 29 ]; then
                    echo "INFO: Sync PBM config from storage again."
                    sync_pbm_config_from_storage
                fi
                sleep $retry_interval
                ((attempt++))
                continue
            else
                echo "ERROR: Failed to get backup metadata: $describe_result"
                exit 1
            fi
        done
        set -e
      
        if [ -z "$describe_result" ] || echo "$describe_result" | grep -q "not found"; then
            echo "ERROR: Failed to get backup metadata after $max_retries attempts"
            exit 1
        fi
      }
      
      function wait_for_restoring() {
        local cnf_file="${MOUNT_DIR}/tmp/pbm_restore.cnf"
        cat <<EOF > ${MOUNT_DIR}/tmp/pbm_restore.cnf
      storage:
        type: s3
        s3:
          region: ${S3_REGION}
          bucket: ${S3_BUCKET}
          prefix: ${S3_PREFIX}
          endpointUrl: ${S3_ENDPOINT}
          forcePathStyle: ${S3_FORCE_PATH_STYLE:-false}
          credentials:
            access-key-id: ${S3_ACCESS_KEY}
            secret-access-key: ${S3_SECRET_KEY}
      EOF
        local attempt=0
        local max_retries=12
        local try_interval=5
        while true; do
          restore_status=$(pbm describe-restore "$restore_name" -c $cnf_file -o json | jq -r '.status')
          echo "INFO: Restore $restore_name status: $restore_status, retrying in ${try_interval}s..."
          if [ "$restore_status" = "done" ]; then
            rm $cnf_file
            break
          elif [ "$restore_status" = "starting" ] || [ "$restore_status" = "running" ]; then
            sleep $try_interval
          elif [ "$restore_status" = "" ]; then
            sleep $try_interval
            attempt=$((attempt+1))
            if [ $attempt -gt $max_retries ]; then
              echo "ERROR: Restore $restore_name status is still empty after $max_retries retries"
              rm $cnf_file
              exit 1
            fi
          else
            rm $cnf_file
            exit 1
          fi
        done
      }
      #!/bin/bash
      # Copyright ApeCloud Co., Ltd. All Rights Reserved.
      # SPDX-License-Identifier: LicenseRef-KubeBlocks-Enterprise
      set -e
      set -o pipefail
      export PATH="$PATH:$DP_DATASAFED_BIN_PATH"
      export DATASAFED_BACKEND_BASE_PATH="$DP_BACKUP_BASE_PATH"
      
      trap handle_exit EXIT
      
      export_pbm_env_vars
      
      set_backup_config_env
      
      export_logs_start_time_env
      
      trap handle_backup_exit EXIT
      
      wait_for_other_operations
      
      sync_pbm_storage_config
      
      wait_for_other_operations
      
      echo "INFO: Starting $PBM_BACKUP_TYPE backup for MongoDB..."
      backup_result=$(pbm backup --type=$PBM_BACKUP_TYPE --mongodb-uri "$PBM_MONGODB_URI" --compression=$PBM_COMPRESSION --wait -o json)
      backup_name=$(echo "$backup_result" | jq -r '.name')
      extras=$(buildJsonString "" "backup_name" "$backup_name")
      extras=$(buildJsonString $extras "backup_type" "$PBM_BACKUP_TYPE")
      echo "INFO: Backup name: $backup_name"
      
      wait_for_backup_completion
      
      echo "INFO: Backup description result:"
      echo "$(echo $describe_result | jq)"
      last_write_time=$(echo "$describe_result" | jq -r '.last_write_time')
      extras=$(buildJsonString $extras "last_write_time" "$last_write_time")
      start_time=$(echo "$describe_result" | jq -r '.name')
      end_unix_time=$(echo "$describe_result" | jq -r '.last_write_ts')
      # used for pitr
      end_time=$(date -u -d "@$((end_unix_time + 2))" +"%Y-%m-%dT%H:%M:%SZ")
      total_size=$(echo "$describe_result" | jq -r '.size')
      DP_save_backup_status_info "$total_size" "$start_time" "$end_time" "" "{$extras}"
      
      print_pbm_logs_by_event "backup"
      
    State:          Terminated
      Reason:       Error
      Exit Code:    1
      Started:      Thu, 13 Aug 2026 15:41:54 +0800
      Finished:     Thu, 13 Aug 2026 15:42:25 +0800
    Ready:          False
    Restart Count:  0
    Limits:
      cpu:                        0
      memory:                     0
      vke.volcengine.com/eni-ip:  1
    Requests:
      cpu:                        0
      memory:                     0
      vke.volcengine.com/eni-ip:  1
    Environment Variables from:
      mongodbs-yccyjr-config-server-env  ConfigMap  Optional: false
    Environment:
      MONGODB_USER:                       <set to the key 'username' in secret 'mongodbs-yccyjr-config-server-account-root'>  Optional: false
      MONGODB_PASSWORD:                   <set to the key 'password' in secret 'mongodbs-yccyjr-config-server-account-root'>  Optional: false
      POD_NAME:                           dp-backup-0-backup-default-mongodbs-yccyjr-20260813154129-zprdb (v1:metadata.name)
      POD_NAMESPACE:                      default (v1:metadata.namespace)
      POD_UID:                             (v1:metadata.uid)
      POD_IP:                              (v1:status.podIP)
      PATH:                               /tools/:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
      KB_SERVICE_CHARACTER_TYPE:          mongodb
      SERVICE_PORT:                       $(KB_SERVICE_PORT)
      MONGODB_ROOT_USER:                  $(MONGODB_USER)
      MONGODB_ROOT_PASSWORD:              $(MONGODB_PASSWORD)
      MONGODB_GRANT_ANYACTION_PRIVILEGE:  true
      PBM_MONGODB_REPLICA_SET:            $(CLUSTER_COMPONENT_NAME)
      PBM_AGENT_SIDECAR:                  true
      PBM_AGENT_SIDECAR_SLEEP:            5
      SHARDED:                            TRUE
      PBM_AGENT_MONGODB_USERNAME:         $(MONGODB_USER)
      PBM_AGENT_MONGODB_PASSWORD:         $(MONGODB_PASSWORD)
      PBM_MONGODB_URI:                    mongodb://$(PBM_AGENT_MONGODB_USERNAME):$(PBM_AGENT_MONGODB_PASSWORD)@localhost:$(KB_SERVICE_PORT)/?authSource=admin
      KB_POD_FQDN:                        $(POD_NAME).$(CLUSTER_COMPONENT_NAME)-headless.$(CLUSTER_NAMESPACE).svc.cluster.local
      MONGODB_BALANCER_ENABLED:           false
      DP_BACKUP_NAME:                     backup-default-mongodbs-yccyjr-20260813154129
      DP_TARGET_POD_NAME:                 mongodbs-yccyjr-config-server-1
      DP_TARGET_POD_ROLE:                 secondary
      DP_BACKUP_BASE_PATH:                /default/mongodbs-yccyjr-3ecff161-5283-4a87-8565-fb4b7fbb1d58/config-server/backup-default-mongodbs-yccyjr-20260813154129
      DP_BACKUP_INFO_FILE:                /dp-manager/backup.info
      DP_TTL:                             
      DP_DB_HOST:                         mongodbs-yccyjr-config-server-1.mongodbs-yccyjr-config-server-headless
      DP_DB_PORT:                         27017
      DP_DB_PASSWORD:                     <set to the key 'password' in secret 'mongodbs-yccyjr-config-server-account-root'>  Optional: false
      DP_DB_USER:                         <set to the key 'username' in secret 'mongodbs-yccyjr-config-server-account-root'>  Optional: false
      DATA_DIR:                           /data/mongodb/db
      MOUNT_DIR:                          /data/mongodb
      PBM_BACKUP_DIR_NAME:                pbm-backups
      PBM_BACKUP_TYPE:                    logical
      PBM_COMPRESSION:                    s2
      PBM_RESTORE_DOWNLOAD_WORKERS:       4
      PBM_IMAGE_TAG:                      2.5.0-multi
      PSM_IMAGE_TAG:                      5.0.29-multi
      KB_CLUSTER_UID:                     3ecff161-5283-4a87-8565-fb4b7fbb1d58
      KB_CLUSTER_NAME:                    mongodbs-yccyjr
      KB_COMP_NAME:                       config-server
      KB_NAMESPACE:                       default
      DP_DATASAFED_BIN_PATH:              /bin/datasafed
    Mounts:
      /bin/datasafed from dp-datasafed-bin (rw)
      /data/mongodb from data (rw)
      /dp-manager from manager-shared-volume (rw)
      /etc/datasafed from dp-datasafed-config (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-wfmhj (ro)
  manager:
    Container ID:  containerd://997f85ebd2a0a34e7cf2d11b8dfdc1c48defbf2c75dc7471545261bcf182dc03
    Image:         apecloud-registry.cn-zhangjiakou.cr.aliyuncs.com/apecloud/kubeblocks-tools:1.1.0-beta.9
    Image ID:      apecloud-registry.cn-zhangjiakou.cr.aliyuncs.com/apecloud/kubeblocks-tools@sha256:d3aede875f92a001d109d432ef054d20a6d0aa68c687fe3564952c054e6e1803
    Port:          <none>
    Host Port:     <none>
    Command:
      sh
      -c
    Args:
      
      set -o errexit
      set -o nounset
      
      export PATH="$PATH:$DP_DATASAFED_BIN_PATH"
      export DATASAFED_BACKEND_BASE_PATH="$DP_BACKUP_BASE_PATH"
      
      backup_info_file="${DP_BACKUP_INFO_FILE}"
      sleep_seconds="${DP_CHECK_INTERVAL}"
      namespace="default"
      backup_name="backup-default-mongodbs-yccyjr-20260813154129"
      
      if [ "$sleep_seconds" -le 0 ]; then
        sleep_seconds=30
      fi
      
      exit_file="${backup_info_file}.exit"
      while true; do
        if [ -f "$exit_file" ]; then
          echo "exit file $exit_file exists, exit"
          exit 1
        fi
        if [ -f "$backup_info_file" ]; then
          break
        fi
        echo "backup info file not exists, wait for ${sleep_seconds}s"
        sleep "$sleep_seconds"
      done
      
      backup_info=$(cat "$backup_info_file")
      echo "backupInfo:${backup_info}"
      
      status="{\"status\":${backup_info}}"
      kubectl -n "$namespace" patch backups.dataprotection.kubeblocks.io "$backup_name" --subresource=status --type=merge --patch "${status}"
      
      # save the backup CR object to the backup repo
      kubectl -n "$namespace" get backups.dataprotection.kubeblocks.io "$backup_name" -o json | datasafed push - "/kubeblocks-backup.json"
      
    State:          Terminated
      Reason:       Error
      Exit Code:    1
      Started:      Thu, 13 Aug 2026 15:41:55 +0800
      Finished:     Thu, 13 Aug 2026 15:42:30 +0800
    Ready:          False
    Restart Count:  0
    Limits:
      cpu:     0
      memory:  0
    Requests:
      cpu:     0
      memory:  0
    Environment Variables from:
      mongodbs-yccyjr-config-server-env  ConfigMap  Optional: false
    Environment:
      MONGODB_USER:                       <set to the key 'username' in secret 'mongodbs-yccyjr-config-server-account-root'>  Optional: false
      MONGODB_PASSWORD:                   <set to the key 'password' in secret 'mongodbs-yccyjr-config-server-account-root'>  Optional: false
      POD_NAME:                           dp-backup-0-backup-default-mongodbs-yccyjr-20260813154129-zprdb (v1:metadata.name)
      POD_NAMESPACE:                      default (v1:metadata.namespace)
      POD_UID:                             (v1:metadata.uid)
      POD_IP:                              (v1:status.podIP)
      PATH:                               /tools/:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
      KB_SERVICE_CHARACTER_TYPE:          mongodb
      SERVICE_PORT:                       $(KB_SERVICE_PORT)
      MONGODB_ROOT_USER:                  $(MONGODB_USER)
      MONGODB_ROOT_PASSWORD:              $(MONGODB_PASSWORD)
      MONGODB_GRANT_ANYACTION_PRIVILEGE:  true
      PBM_MONGODB_REPLICA_SET:            $(CLUSTER_COMPONENT_NAME)
      PBM_AGENT_SIDECAR:                  true
      PBM_AGENT_SIDECAR_SLEEP:            5
      SHARDED:                            TRUE
      PBM_AGENT_MONGODB_USERNAME:         $(MONGODB_USER)
      PBM_AGENT_MONGODB_PASSWORD:         $(MONGODB_PASSWORD)
      PBM_MONGODB_URI:                    mongodb://$(PBM_AGENT_MONGODB_USERNAME):$(PBM_AGENT_MONGODB_PASSWORD)@localhost:$(KB_SERVICE_PORT)/?authSource=admin
      KB_POD_FQDN:                        $(POD_NAME).$(CLUSTER_COMPONENT_NAME)-headless.$(CLUSTER_NAMESPACE).svc.cluster.local
      MONGODB_BALANCER_ENABLED:           false
      DP_BACKUP_NAME:                     backup-default-mongodbs-yccyjr-20260813154129
      DP_TARGET_POD_NAME:                 mongodbs-yccyjr-config-server-1
      DP_TARGET_POD_ROLE:                 secondary
      DP_BACKUP_BASE_PATH:                /default/mongodbs-yccyjr-3ecff161-5283-4a87-8565-fb4b7fbb1d58/config-server/backup-default-mongodbs-yccyjr-20260813154129
      DP_BACKUP_INFO_FILE:                /dp-manager/backup.info
      DP_TTL:                             
      DP_DB_HOST:                         mongodbs-yccyjr-config-server-1.mongodbs-yccyjr-config-server-headless
      DP_DB_PORT:                         27017
      DP_DB_PASSWORD:                     <set to the key 'password' in secret 'mongodbs-yccyjr-config-server-account-root'>  Optional: false
      DP_DB_USER:                         <set to the key 'username' in secret 'mongodbs-yccyjr-config-server-account-root'>  Optional: false
      DATA_DIR:                           /data/mongodb/db
      MOUNT_DIR:                          /data/mongodb
      PBM_BACKUP_DIR_NAME:                pbm-backups
      PBM_BACKUP_TYPE:                    logical
      PBM_COMPRESSION:                    s2
      PBM_RESTORE_DOWNLOAD_WORKERS:       4
      PBM_IMAGE_TAG:                      2.5.0-multi
      PSM_IMAGE_TAG:                      5.0.29-multi
      KB_CLUSTER_UID:                     3ecff161-5283-4a87-8565-fb4b7fbb1d58
      KB_CLUSTER_NAME:                    mongodbs-yccyjr
      KB_COMP_NAME:                       config-server
      KB_NAMESPACE:                       default
      DP_DATASAFED_BIN_PATH:              /bin/datasafed
      DP_CHECK_INTERVAL:                  5
    Mounts:
      /bin/datasafed from dp-datasafed-bin (rw)
      /data/mongodb from data (rw)
      /dp-manager from manager-shared-volume (rw)
      /etc/datasafed from dp-datasafed-config (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-wfmhj (ro)
Conditions:
  Type                        Status
  PodReadyToStartContainers   False 
  Initialized                 True 
  Ready                       False 
  ContainersReady             False 
  PodScheduled                True 
Volumes:
  manager-shared-volume:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     
    SizeLimit:  <unset>
  data:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  data-mongodbs-yccyjr-config-server-1
    ReadOnly:   false
  dp-datasafed-config:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  tool-config-backuprepo-kbcli-test-bnfj2s
    Optional:    false
  dp-datasafed-bin:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     
    SizeLimit:  <unset>
  kube-api-access-wfmhj:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   BestEffort
Node-Selectors:              kubernetes.io/hostname=192.168.0.134
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type    Reason     Age    From               Message
  ----    ------     ----   ----               -------
  Normal  Scheduled  4m44s  default-scheduler  Successfully assigned default/dp-backup-0-backup-default-mongodbs-yccyjr-20260813154129-zprdb to 192.168.0.134
  Normal  Pulled     4m44s  kubelet            Container image "apecloud-registry.cn-zhangjiakou.cr.aliyuncs.com/apecloud/datasafed:0.2.3" already present on machine
  Normal  Created    4m44s  kubelet            Created container dp-copy-datasafed
  Normal  Started    4m44s  kubelet            Started container dp-copy-datasafed
  Normal  Pulled     4m43s  kubelet            Container image "apecloud-registry.cn-zhangjiakou.cr.aliyuncs.com/apecloud/percona-backup-mongodb:2.5.0-multi" already present on machine
  Normal  Created    4m43s  kubelet            Created container backupdata
  Normal  Started    4m43s  kubelet            Started container backupdata
  Normal  Pulled     4m43s  kubelet            Container image "apecloud-registry.cn-zhangjiakou.cr.aliyuncs.com/apecloud/kubeblocks-tools:1.1.0-beta.9" already present on machine
  Normal  Created    4m42s  kubelet            Created container manager
  Normal  Started    4m42s  kubelet            Started container manager
```
logs error pod
```
➜  ~ kubectl logs dp-backup-0-backup-default-mongodbs-yccyjr-20260813154129-zprdb
Defaulted container "backupdata" out of: backupdata, manager, dp-copy-datasafed (init)
2026-08-13 07:41:54 INFO: storage config have been extracted.
INFO: Other operation backup is running, waiting... (1/60)
INFO: Other operation backup is running, waiting... (2/60)
INFO: Other operation backup is running, waiting... (3/60)
INFO: Other operation backup is running, waiting... (4/60)
INFO: Checking if PBM storage config exists
INFO: Current PBM storage endpoint: http://kbcli-test-minio.kb-system.svc.cluster.local:9000
INFO: Current PBM storage region: =
INFO: Current PBM storage bucket: kbcli-test
INFO: Current PBM storage prefix: default/mongodbs-yccyjr-3ecff161-5283-4a87-8565-fb4b7fbb1d58/config-server/pbm-backups
INFO: PBM storage config already exists.
INFO: Starting logical backup for MongoDB...
INFO: Backup name: 2026-08-13T07:42:15Z
INFO: Backup metadata not found, retrying in 5s...
INFO: Backup status is running, retrying in 5s...
ERROR: Backup failed with status: dumpDone
INFO: Printing PBM tail logs
2026-08-13T07:42:15Z I [mongodbs-yccyjr-shard-f47/mongodbs-yccyjr-shard-f47-2.mongodbs-yccyjr-shard-f47-headless.default.svc.cluster.local:27017] got epoch {1786606901 31}
2026-08-13T07:42:15Z I [mongodbs-yccyjr-shard-f47/mongodbs-yccyjr-shard-f47-1.mongodbs-yccyjr-shard-f47-headless.default.svc.cluster.local:27017] got command backup [name: 2026-08-13T07:42:15Z, compression: s2 (level: default)] <ts: 1786606935>
2026-08-13T07:42:15Z I [mongodbs-yccyjr-shard-f47/mongodbs-yccyjr-shard-f47-1.mongodbs-yccyjr-shard-f47-headless.default.svc.cluster.local:27017] got epoch {1786606901 31}
2026-08-13T07:42:15Z I [mongodbs-yccyjr-shard-f47/mongodbs-yccyjr-shard-f47-0.mongodbs-yccyjr-shard-f47-headless.default.svc.cluster.local:27017] got command backup [name: 2026-08-13T07:42:15Z, compression: s2 (level: default)] <ts: 1786606935>
2026-08-13T07:42:15Z I [mongodbs-yccyjr-shard-f47/mongodbs-yccyjr-shard-f47-0.mongodbs-yccyjr-shard-f47-headless.default.svc.cluster.local:27017] got epoch {1786606901 31}
2026-08-13T07:42:15Z I [mongodbs-yccyjr-shard-nsw/mongodbs-yccyjr-shard-nsw-2.mongodbs-yccyjr-shard-nsw-headless.default.svc.cluster.local:27017] got command backup [name: 2026-08-13T07:42:15Z, compression: s2 (level: default)] <ts: 1786606935>
2026-08-13T07:42:15Z I [mongodbs-yccyjr-shard-nsw/mongodbs-yccyjr-shard-nsw-2.mongodbs-yccyjr-shard-nsw-headless.default.svc.cluster.local:27017] got epoch {1786606901 31}
2026-08-13T07:42:16Z I [mongodbs-yccyjr-shard-nsw/mongodbs-yccyjr-shard-nsw-0.mongodbs-yccyjr-shard-nsw-headless.default.svc.cluster.local:27017] got command backup [name: 2026-08-13T07:42:15Z, compression: s2 (level: default)] <ts: 1786606935>
2026-08-13T07:42:16Z I [mongodbs-yccyjr-shard-nsw/mongodbs-yccyjr-shard-nsw-0.mongodbs-yccyjr-shard-nsw-headless.default.svc.cluster.local:27017] got epoch {1786606901 31}
2026-08-13T07:42:16Z I [mongodbs-yccyjr-shard-nsw/mongodbs-yccyjr-shard-nsw-1.mongodbs-yccyjr-shard-nsw-headless.default.svc.cluster.local:27017] got command backup [name: 2026-08-13T07:42:15Z, compression: s2 (level: default)] <ts: 1786606935>
2026-08-13T07:42:16Z I [mongodbs-yccyjr-shard-nsw/mongodbs-yccyjr-shard-nsw-1.mongodbs-yccyjr-shard-nsw-headless.default.svc.cluster.local:27017] got epoch {1786606901 31}
2026-08-13T07:42:16Z I [mongodbs-yccyjr-shard-v49/mongodbs-yccyjr-shard-v49-1.mongodbs-yccyjr-shard-v49-headless.default.svc.cluster.local:27017] [backup/2026-08-13T07:42:15Z] backup started
2026-08-13T07:42:16Z I [mongodbs-yccyjr-config-server/mongodbs-yccyjr-config-server-2.mongodbs-yccyjr-config-server-headless.default.svc.cluster.local:27017] [backup/2026-08-13T07:42:15Z] backup started
2026-08-13T07:42:16Z I [mongodbs-yccyjr-shard-f47/mongodbs-yccyjr-shard-f47-2.mongodbs-yccyjr-shard-f47-headless.default.svc.cluster.local:27017] [backup/2026-08-13T07:42:15Z] backup started
2026-08-13T07:42:16Z I [mongodbs-yccyjr-shard-nsw/mongodbs-yccyjr-shard-nsw-2.mongodbs-yccyjr-shard-nsw-headless.default.svc.cluster.local:27017] [backup/2026-08-13T07:42:15Z] backup started
2026-08-13T07:42:19Z I [mongodbs-yccyjr-shard-v49/mongodbs-yccyjr-shard-v49-1.mongodbs-yccyjr-shard-v49-headless.default.svc.cluster.local:27017] [backup/2026-08-13T07:42:15Z] mongodump finished, waiting for the oplog
2026-08-13T07:42:19Z I [mongodbs-yccyjr-shard-f47/mongodbs-yccyjr-shard-f47-2.mongodbs-yccyjr-shard-f47-headless.default.svc.cluster.local:27017] [backup/2026-08-13T07:42:15Z] mongodump finished, waiting for the oplog
2026-08-13T07:42:19Z I [mongodbs-yccyjr-shard-nsw/mongodbs-yccyjr-shard-nsw-2.mongodbs-yccyjr-shard-nsw-headless.default.svc.cluster.local:27017] [backup/2026-08-13T07:42:15Z] mongodump finished, waiting for the oplog
2026-08-13T07:42:20Z I [mongodbs-yccyjr-config-server/mongodbs-yccyjr-config-server-2.mongodbs-yccyjr-config-server-headless.default.svc.cluster.local:27017] [backup/2026-08-13T07:42:15Z] mongodump finished, waiting for the oplog
2026-08-13T07:42:22Z I [mongodbs-yccyjr-config-server/mongodbs-yccyjr-config-server-2.mongodbs-yccyjr-config-server-headless.default.svc.cluster.local:27017] [backup/2026-08-13T07:42:15Z] created chunk 2026-08-13T07:42:16 - 2026-08-13T07:42:21
failed with exit code 1
```


**Expected behavior**
A clear and concise description of what you expected to happen.

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Desktop (please complete the following information):**
 - OS: [e.g. iOS]
 - Browser [e.g. chrome, safari]
 - Version [e.g. 22]

**Additional context**
Add any other context about the problem here.

--- a/addons/mongodb/dataprotection/common-scripts.sh
+++ b/addons/mongodb/dataprotection/common-scripts.sh
@@ -379,25 +379,31 @@ function sync_pbm_config_from_storage() {
 function wait_for_backup_completion() {
   describe_result=""
   local retry_interval=5
-  local attempt=1
+  local attempt=0
   local max_retries=12
   set +e
   while true; do
     describe_result=$(pbm describe-backup --mongodb-uri "$PBM_MONGODB_URI" "$backup_name" -o json 2>&1)
     if [ $? -eq 0 ] && [ -n "$describe_result" ]; then
       backup_status=$(echo "$describe_result" | jq -r '.status')
-      if [ "$backup_status" = "starting" ] || [ "$backup_status" = "running" ]; then
-        echo "INFO: Backup status is $backup_status, retrying in ${retry_interval}s..."
-      elif [ "$backup_status" = "" ]; then
-        echo "INFO: Backup status is $backup_status, retrying in ${retry_interval}s..."
-        attempt=$((attempt+1))
-      elif [ "$backup_status" = "done" ]; then
-        echo "INFO: Backup status is done."
-        break
-      else
-        echo "ERROR: Backup failed with status: $backup_status"
-        exit 1
-      fi
+      case "$backup_status" in
+        done)
+          echo "INFO: Backup status is done."
+          break
+          ;;
+        error|canceled)
+          echo "ERROR: Backup failed with status: $backup_status"
+          exit 1
+          ;;
+        "")
+          echo "INFO: Backup status is empty, retrying in ${retry_interval}s..."
+          attempt=$((attempt+1))
+          ;;
+        *)
+          # PBM treats every non-empty status except done, canceled, and error as running.
+          echo "INFO: Backup status is $backup_status, retrying in ${retry_interval}s..."
+          ;;
+      esac
     elif echo "$describe_result" | grep -q "not found"; then
       echo "INFO: Backup metadata not found, retrying in ${retry_interval}s..."
       attempt=$((attempt+1))
@@ -405,11 +411,11 @@ function wait_for_backup_completion() {
       echo "ERROR: Unexpected: $describe_result"
       exit 1
     fi
-    sleep $retry_interval
-    if [ $attempt -gt $max_retries ]; then
+    if [ $attempt -ge $max_retries ]; then
       echo "ERROR: Failed to get backup status after $max_retries attempts"
       exit 1
     fi
+    sleep $retry_interval
   done
   set -e
 
@@ -535,8 +541,15 @@ function get_describe_backup_info() {
 }
 
 function wait_for_restoring() {
-  local cnf_file="${MOUNT_DIR}/tmp/pbm_restore.cnf"
-  cat <<EOF > ${MOUNT_DIR}/tmp/pbm_restore.cnf
+  local cnf_file=""
+  local describe_result=""
+  local restore_status=""
+  case "${PBM_BACKUP_TYPE:-}" in
+    logical)
+      ;;
+    physical|continuous)
+      cnf_file="${MOUNT_DIR}/tmp/pbm_restore.cnf"
+      cat <<EOF > "$cnf_file"
 storage:
   type: s3
   s3:
@@ -549,28 +562,64 @@ storage:
       access-key-id: ${S3_ACCESS_KEY}
       secret-access-key: ${S3_SECRET_KEY}
 EOF
+      ;;
+    *)
+      echo "ERROR: Unsupported PBM backup type: ${PBM_BACKUP_TYPE:-<empty>}"
+      exit 1
+      ;;
+  esac
   local attempt=0
   local max_retries=12
   local try_interval=5
   while true; do
-    restore_status=$(pbm describe-restore "$restore_name" -c $cnf_file -o json | jq -r '.status')
-    echo "INFO: Restore $restore_name status: $restore_status, retrying in ${try_interval}s..."
-    if [ "$restore_status" = "done" ]; then
-      rm $cnf_file
-      break
-    elif [ "$restore_status" = "starting" ] || [ "$restore_status" = "running" ]; then
-      sleep $try_interval
-    elif [ "$restore_status" = "" ]; then
-      sleep $try_interval
-      attempt=$((attempt+1))
-      if [ $attempt -gt $max_retries ]; then
-        echo "ERROR: Restore $restore_name status is still empty after $max_retries retries"
-        rm $cnf_file
+    if [ "$PBM_BACKUP_TYPE" = "logical" ]; then
+      if describe_result=$(pbm describe-restore "$restore_name" --mongodb-uri "$PBM_MONGODB_URI" -o json 2>&1); then
+        :
+      elif echo "$describe_result" | grep -Eq "not found|no documents in result|undefined restore meta"; then
+        describe_result="{}"
+      else
+        echo "ERROR: Failed to get restore status: $describe_result"
         exit 1
       fi
     else
-      rm $cnf_file
+      if ! describe_result=$(pbm describe-restore "$restore_name" -c "$cnf_file" -o json 2>&1); then
+        echo "ERROR: Failed to get restore status: $describe_result"
+        rm -f "$cnf_file"
+        exit 1
+      fi
+    fi
+
+    if ! restore_status=$(echo "$describe_result" | jq -r '.status // empty'); then
+      echo "ERROR: Failed to parse restore status: $describe_result"
+      [ -z "$cnf_file" ] || rm -f "$cnf_file"
       exit 1
     fi
+
+    case "$restore_status" in
+      done)
+        echo "INFO: Restore $restore_name status: done."
+        [ -z "$cnf_file" ] || rm -f "$cnf_file"
+        break
+        ;;
+      error|canceled)
+        echo "ERROR: Restore $restore_name failed with status: $restore_status"
+        [ -z "$cnf_file" ] || rm -f "$cnf_file"
+        exit 1
+        ;;
+      "")
+        attempt=$((attempt+1))
+        if [ $attempt -ge $max_retries ]; then
+          echo "ERROR: Restore $restore_name status is still empty after $max_retries attempts"
+          [ -z "$cnf_file" ] || rm -f "$cnf_file"
+          exit 1
+        fi
+        echo "INFO: Restore $restore_name status is empty, retrying in ${try_interval}s..."
+        ;;
+      *)
+        # PBM treats every non-empty status except done, canceled, and error as running.
+        echo "INFO: Restore $restore_name status: $restore_status, retrying in ${try_interval}s..."
+        ;;
+    esac
+    sleep $try_interval
   done
 }

--- a/addons/mongodb/scripts/mongodb-shard-manage.sh
+++ b/addons/mongodb/scripts/mongodb-shard-manage.sh
@@ -31,8 +31,7 @@ wait_for_mongos() {
 check_shard_exists() {
     # check if the shard exists in the config database
     local shard_exists
-    shard_exists=$($CLUSTER_MONGO "db.getSiblingDB(\"config\").shards.find({ _id: \"$MONGODB_REPLICA_SET_NAME\" })" 2>/dev/null)
-    if [ $? -ne 0 ]; then
+    if ! shard_exists=$($CLUSTER_MONGO "db.getSiblingDB(\"config\").shards.find({ _id: \"$MONGODB_REPLICA_SET_NAME\" })" 2>/dev/null); then
         echo "ERROR: Failed to check if shard $MONGODB_REPLICA_SET_NAME exists." >&2
         exit 1
     fi
@@ -69,9 +68,9 @@ get_remove_shard_status() {
     # Execute the removeShard command and capture its JSON output
     local result
     if [ "$CLIENT" = "mongosh" ]; then
-        result=$($CLUSTER_MONGO "EJSON.stringify(db.adminCommand( { removeShard: \"$MONGODB_REPLICA_SET_NAME\" } ))")
+        result=$($CLUSTER_MONGO "EJSON.stringify(db.adminCommand( { removeShard: \"$MONGODB_REPLICA_SET_NAME\" } ))") || return 1
     else
-        result=$($CLUSTER_MONGO "JSON.stringify(db.adminCommand( { removeShard: \"$MONGODB_REPLICA_SET_NAME\" } ))")
+        result=$($CLUSTER_MONGO "JSON.stringify(db.adminCommand( { removeShard: \"$MONGODB_REPLICA_SET_NAME\" } ))") || return 1
     fi
     echo "$result"
 }
@@ -80,7 +79,7 @@ get_remove_shard_state() {
     local result=$1
     # Parse and log the state using jq
     local state
-    state=$(echo "$result" | jq -r '.state')
+    state=$(echo "$result" | jq -er '.state // empty') || return 1
     # Return the state as the function output
     echo "$state"
 }
@@ -90,9 +89,9 @@ get_remaining_jumbo_chunks() {
     # Parse and log the jumboChunks count using jq
     local jumbo_chunks
     if [ "$CLIENT" = "mongosh" ]; then
-        jumbo_chunks=$(echo "$result" | jq -r '.remaining.jumboChunks // 0')
+        jumbo_chunks=$(echo "$result" | jq -er '(.remaining.jumboChunks // 0) | tonumber') || return 1
     else
-        jumbo_chunks=$(echo "$result" | jq -r '.remaining.jumboChunks.numberLong // 0')
+        jumbo_chunks=$(echo "$result" | jq -er '(.remaining.jumboChunks."$numberLong" // 0) | tonumber') || return 1
     fi
     # Return the jumboChunks count as the function output
     echo "$jumbo_chunks"
@@ -103,12 +102,31 @@ get_remaining_chunks() {
     # Parse and log the chunks count using jq
     local chunks
     if [ "$CLIENT" = "mongosh" ]; then
-        chunks=$(echo "$result" | jq -r '.remaining.chunks // 0')
+        chunks=$(echo "$result" | jq -er '(.remaining.chunks // 0) | tonumber') || return 1
     else
-        chunks=$(echo "$result" | jq -r '.remaining.chunks.numberLong // 0')
+        chunks=$(echo "$result" | jq -er '(.remaining.chunks."$numberLong" // 0) | tonumber') || return 1
     fi
     # Return the chunks count as the function output
     echo "$chunks"
+}
+
+get_database_primary() {
+    local database=$1
+    local result
+    result=$($CLUSTER_MONGO "JSON.stringify(db.getSiblingDB('config').databases.findOne({ _id: \"$database\" }))") || return 1
+    echo "$result" | jq -r '.primary // empty'
+}
+
+get_destination_shard() {
+    local result
+    local shards
+    result=$($CLUSTER_MONGO "JSON.stringify(
+        db.getSiblingDB('config').shards.find({
+            _id: { \$ne: '$MONGODB_REPLICA_SET_NAME' }
+        }).toArray()
+    )") || return 1
+    shards=$(echo "$result" | jq -r '.[]._id') || return 1
+    echo "$shards" | shuf -n 1
 }
 
 delete_or_scale_in_mongodb_shard() {
@@ -116,73 +134,104 @@ delete_or_scale_in_mongodb_shard() {
 
     if ! check_shard_exists; then
         echo "INFO: Shard $MONGODB_REPLICA_SET_NAME does not exist, skipping scale-in."
-        exit 0
+        return 0
     fi
 
-    original_balance_status=$($CLUSTER_MONGO "sh.getBalancerState()")
-    if [ "$original_balance_status" = "false" ]; then
-        $CLUSTER_MONGO "sh.startBalancer()"
-    fi
+    balancer_status=$($CLUSTER_MONGO "sh.getBalancerState()") || {
+        echo "ERROR: Failed to get the balancer state." >&2
+        return 1
+    }
+    case "$balancer_status" in
+    false)
+        if ! $CLUSTER_MONGO "sh.startBalancer()"; then
+            echo "ERROR: Failed to start the balancer." >&2
+            return 1
+        fi
+        ;;
+    true)
+        ;;
+    *)
+        echo "ERROR: Unexpected balancer state: $balancer_status" >&2
+        return 1
+        ;;
+    esac
 
     echo "INFO: Shard $MONGODB_REPLICA_SET_NAME exists, scaling in..."
-    # Remove the shard and wait until the state is 'completed'
-    while true; do
-
-        if ! check_shard_exists; then
-            echo "INFO: Shard $MONGODB_REPLICA_SET_NAME does not exist, exiting."
-            exit 0
-        fi
-
-        status_json=$(get_remove_shard_status)
-        echo "INFO: Remove shard status: $status_json"
-        state=$(get_remove_shard_state "$status_json")
-        echo "INFO: Current state of shard $MONGODB_REPLICA_SET_NAME is $state"
-        if [ "$state" = "completed" ]; then
-            break
-        elif [ "$state" = "ongoing" ]; then
-            remaining_jumboChunks=$(get_remaining_jumbo_chunks "$status_json")
-            if [ "$remaining_jumboChunks" -gt 0 ]; then
-                echo "INFO: $remaining_jumboChunks jumbo chunks remaining, please clear jumbo chunks before removing the shard."
-                exit 1
-            fi
-
-            remaining_chunks=$(get_remaining_chunks "$status_json")
-            echo "INFO: $remaining_chunks chunks remaining."
-            if [ "$remaining_chunks" -eq 0 ]; then
-                dbs_to_move=$(echo "$status_json" | jq -r '.dbsToMove[]')
-                note=$(echo "$status_json" | jq -r '.note')
-                echo "INFO: $note"
-                echo "$dbs_to_move"
-                for db in $dbs_to_move; do
-                    echo "INFO: Database '$db' is scheduled for movePrimary..."
-                    if [ -z "$DESTINATION_SHARD" ]; then
-                        DESTINATION_SHARD=$($CLUSTER_MONGO "JSON.stringify(
-                            db.getSiblingDB('config').shards.find({
-                                _id: { \$ne: '$MONGODB_REPLICA_SET_NAME' }
-                            }).toArray()
-                        )" | jq -r '.[]._id' | shuf -n 1)
-                        if [ -z "$DESTINATION_SHARD" ]; then
-                            echo "ERROR: No available shard found for moving primary for database '$db'."
-                            exit 1
-                        fi
-                    fi
-                    echo "INFO: Moving primary for database '$db' to shard '$DESTINATION_SHARD'..."
-                    $CLUSTER_MONGO "db.adminCommand({ movePrimary: \"$db\", to: \"$DESTINATION_SHARD\" })"
-                done
-                continue
-            else
-                echo "INFO: $remaining_chunks chunks are still being migrated, waiting..."
-            fi
-        fi
-        sleep 2
-    done
-
-    # reset balancer state
-    if [ "$original_balance_status" = "false" ]; then
-        $CLUSTER_MONGO "sh.stopBalancer()"
-        echo "INFO: Balancer state has been reset to false."
+    # orphanCleanupDelaySecs controls how long MongoDB delays cleaning orphaned
+    # documents after a chunk migration. removeShard may wait for this delay
+    # (900 seconds by default) before the same range can be migrated again.
+    status_json=$(get_remove_shard_status) || {
+        echo "ERROR: Failed to get the remove shard status." >&2
+        return 1
+    }
+    echo "INFO: Remove shard status: $status_json"
+    state=$(get_remove_shard_state "$status_json") || {
+        echo "ERROR: Failed to parse the remove shard state." >&2
+        return 1
+    }
+    echo "INFO: Current state of shard $MONGODB_REPLICA_SET_NAME is $state"
+    if [ "$state" = "completed" ]; then
+        echo "INFO: Shard $MONGODB_REPLICA_SET_NAME has been successfully removed."
+        return 0
     fi
-    echo "INFO: Shard $MONGODB_REPLICA_SET_NAME has been successfully removed."
+
+    if [ "$state" = "ongoing" ]; then
+        remaining_jumbo_chunks=$(get_remaining_jumbo_chunks "$status_json") || {
+            echo "ERROR: Failed to parse the remaining jumbo chunks." >&2
+            return 1
+        }
+        if [ "$remaining_jumbo_chunks" -gt 0 ]; then
+            echo "ERROR: $remaining_jumbo_chunks jumbo chunks remain; clear them before removing the shard." >&2
+            return 1
+        fi
+
+        remaining_chunks=$(get_remaining_chunks "$status_json") || {
+            echo "ERROR: Failed to parse the remaining chunks." >&2
+            return 1
+        }
+        echo "INFO: $remaining_chunks chunks remaining."
+        if [ "$remaining_chunks" -eq 0 ]; then
+            dbs_to_move=$(echo "$status_json" | jq -r '.dbsToMove[]?') || {
+                echo "ERROR: Failed to parse the databases to move." >&2
+                return 1
+            }
+            note=$(echo "$status_json" | jq -r '.note // empty') || {
+                echo "ERROR: Failed to parse the remove shard note." >&2
+                return 1
+            }
+            [ -n "$note" ] && echo "INFO: $note"
+            for db in $dbs_to_move; do
+                current_primary=$(get_database_primary "$db") || {
+                    echo "ERROR: Failed to get the primary shard for database '$db'." >&2
+                    return 1
+                }
+                if [ "$current_primary" != "$MONGODB_REPLICA_SET_NAME" ]; then
+                    echo "INFO: Database '$db' no longer uses shard '$MONGODB_REPLICA_SET_NAME' as primary, skipping."
+                    continue
+                fi
+
+                if [ -z "$DESTINATION_SHARD" ]; then
+                    DESTINATION_SHARD=$(get_destination_shard) || {
+                        echo "ERROR: Failed to get a destination shard for database '$db'." >&2
+                        return 1
+                    }
+                fi
+                if [ -z "$DESTINATION_SHARD" ]; then
+                    echo "ERROR: No available shard found for moving primary for database '$db'." >&2
+                    return 1
+                fi
+
+                echo "INFO: Moving primary for database '$db' to shard '$DESTINATION_SHARD'..."
+                if ! $CLUSTER_MONGO "db.adminCommand({ movePrimary: \"$db\", to: \"$DESTINATION_SHARD\" })"; then
+                    echo "ERROR: Failed to move primary for database '$db'." >&2
+                    return 1
+                fi
+            done
+        fi
+    fi
+
+    echo "ERROR: Shard $MONGODB_REPLICA_SET_NAME has not been removed; the controller will retry." >&2
+    return 1
 }
 
 # main
@@ -202,7 +251,7 @@ if [ $# -eq 1 ]; then
     ;;
   --pre-terminate)
     delete_or_scale_in_mongodb_shard
-    exit 0
+    exit $?
     ;;
   *)
     echo "Error: invalid option '$1'"

--- a/addons/mongodb/templates/shardingdefinition.yaml
+++ b/addons/mongodb/templates/shardingdefinition.yaml
@@ -16,6 +16,8 @@ spec:
   updateStrategy: Parallel
   lifecycleActions:
     shardRemove:
+      # Each attempt is idempotent. Use the controller's default action timeout
+      # and rate limiter to retry removeShard without blocking reconciliation.
       exec:
         container: mongodb
         command:

